### PR TITLE
fix: keep per-wave shutdown floors within coordinator margin

### LINKED ISSUE #1809: Keep per-wave shutdown floors within the coordinator margin
## Description

The shutdown budget reserves `COORDINATOR_MARGIN_FRACTION` (10%) of the total as the gap between the body's own deadline and the coordinator's outer backstop (`resources/lifecycle.py:36-41`). But `Hassette._shutdown_children` (`core/core.py:753-800`) gives **every** wave `max(children_floor_seconds, body_deadline - now)` via `children_budget_remaining()` — the 1s floor applies per wave even after `body_deadline` has passed.

With W waves whose children all hang, the body overruns `body_deadline` by up to W×1s while the margin is only `0.1 × total`. At the 30s default (margin 3s) and ~7 shutdown waves, 4+ hung waves guarantee the coordinator's outer wait fires first: `SHUTDOWN_BODY_TIMED_OUT` + `_force_terminal()` — the crude backstop — becomes the *normal* outcome in exactly the multi-wave-hang scenario the graceful wave-by-wave logic (force-terminate the timed-out wave, still give later waves their chance to close OS resources) was designed for. The graceful path is mathematically unreachable there.

When it fires, the outcome is still safe (terminal state + evidence recorded), so this is priority:low — but it silently defeats the budget machinery's design intent.

## Decision (2026-09-14)

Adaptive per-wave division (`max(small_floor, remaining_time / waves_left)`, recomputed each iteration from the actual runtime wave count) — not a static `margin ≥ expected_waves × floor` config validation. `Hassette._shutdown_children()` already has the full wave list (`self._init_waves`) at shutdown time, so "waves left" is a real runtime value, not something that needs a hardcoded "expected waves" constant that would silently drift as the dependency graph evolves.

**Scoping note:** `children_budget_remaining()` has two callers with genuinely different shapes:
- `Resource._shutdown_children()` (`resources/base.py:416`) — the generic base version. Shuts down **all** children in a single `shutdown_batch()` call, no waves. This one is not affected by the bug at all (one call, one floor) and must keep its exact current behavior.
- `Hassette._shutdown_children()` (`core/core.py`) — the root's dependency-ordered **multi-wave** loop. This is the only affected caller.

Fix: add an optional `waves_left` parameter to `children_budget_remaining()`, defaulting to `1` so `Resource._shutdown_children()`'s existing single-call behavior is byte-for-byte unchanged. `Hassette._shutdown_children()` passes the actual remaining non-empty wave count, computed from `self._init_waves`, decrementing each iteration.

New small per-wave floor for the multi-wave case: `CHILDREN_WAVE_FLOOR_SECONDS = 0.1` (vs. the existing `CHILDREN_FLOOR_SECONDS = 1.0`, kept as-is for the single-batch case). Sanity-checked against the numbers in this issue: at the ~7-wave, 30s-default case cited above, worst-case total floor overrun becomes `7 × 0.1s = 0.7s`, comfortably inside the `3s` margin — versus the current `7 × 1.0s = 7s`, which is exactly what blows past the margin today.

## Fix Shape

**`resources/lifecycle.py`**:

```python
CHILDREN_WAVE_FLOOR_SECONDS = 0.1
"""Minimum guaranteed budget per wave when Hassette._shutdown_children() divides the
remaining body time across multiple dependency-ordered waves (see children_budget_remaining()'s
waves_left parameter). Deliberately much smaller than CHILDREN_FLOOR_SECONDS -- that floor is
sized for a single shutdown_children() call; applied per-wave across every wave in a multi-wave
shutdown, it can blow past COORDINATOR_MARGIN_FRACTION's margin on its own. See #1809."""


def children_budget_remaining(resource: _LifecycleHostP, *, waves_left: int = 1) -> float:
    """Seconds available for the next _shutdown_children() wave.

    Children run last and benefit from any slack the earlier stages left behind. Returns at
    least children_floor_seconds when called once (waves_left=1, the default -- matches
    Resource._shutdown_children()'s single shutdown_batch() call) even when the body is over
    budget.

    When more than one wave shares this budget (Hassette._shutdown_children()'s
    dependency-ordered waves), pass the actual remaining wave count via waves_left so the
    remaining time is divided across waves instead of each wave independently claiming a fresh
    floor -- otherwise W hung waves could overrun body_deadline by up to W times the floor,
    exceeding the coordinator's own margin (see #1809).
    """
    resource = typing.cast("LifecycleMixin", resource)
    budget = resource._shutdown_budget
    if budget is None:
        return resource.hassette.config.lifecycle.resource_shutdown_timeout_seconds
    remaining = budget.body_deadline - asyncio.get_running_loop().time()
    if waves_left <= 1:
        return max(budget.children_floor_seconds, remaining)
    return max(CHILDREN_WAVE_FLOOR_SECONDS, remaining / waves_left)
```

**`core/core.py`** — `Hassette._shutdown_children()`:

```python
type_to_instance = {type(c): c for c in self.children}

child_reports: list[TeardownReport] = []
causes: list[TeardownCause] = []
affected: list[str] = []

non_empty_waves = [wt for wt in reversed(self._init_waves) if any(t in type_to_instance for t in wt)]
waves_left = len(non_empty_waves)

waves_start = asyncio.get_running_loop().time()
for wave_types in non_empty_waves:
    wave = [type_to_instance[t] for t in wave_types if t in type_to_instance]
    timeout = min(
        self.config.lifecycle.resource_shutdown_timeout_seconds,
        children_budget_remaining(self, waves_left=waves_left),
    )
    self.logger.debug("Shutting down wave: [%s]", ", ".join(c.class_name for c in wave))
    wave_start = asyncio.get_running_loop().time()
    result = await shutdown_batch(self, wave, timeout)
    self.logger.debug(
        "Shutdown wave [%s] completed in %.2fs",
        ", ".join(c.class_name for c in wave),
        elapsed_since(wave_start),
    )
    child_reports.extend(result.reports)
    causes.extend(result.causes)
    affected.extend(result.affected)
    waves_left -= 1

self.logger.debug("All shutdown waves completed in %.2fs", elapsed_since(waves_start))
return finalize_shutdown_report(child_reports, causes, affected)
```

(The `if not wave: continue` in the original loop is subsumed by pre-filtering into `non_empty_waves`.)

## Key Items

- `src/hassette/resources/lifecycle.py:52-65` — add `CHILDREN_WAVE_FLOOR_SECONDS` constant.
- `src/hassette/resources/lifecycle.py:570-580` — `children_budget_remaining()` gains `waves_left` parameter.
- `src/hassette/core/core.py:754-801` — `Hassette._shutdown_children()` pre-filters non-empty waves, tracks `waves_left`, passes it through.

## Sequencing note

#2066 (already decided, quality-scanner) does a pure mechanical dedup of the shared preamble across these same `lifecycle.py` budget-accessor functions — doesn't touch `children_budget_remaining()`'s formula, so no conflict, just rebase whichever of the two PRs lands second.

## Acceptance Criteria

- [ ] The sum of per-wave floors cannot exceed the coordinator margin at realistic wave counts (verified: 7 waves × 0.1s = 0.7s vs. 3s margin at 30s default).
- [ ] Multi-wave-hang test: with all waves hanging, the body still finishes (with per-wave force-terminals and evidence) before the coordinator's outer `total_deadline`, i.e. no `SHUTDOWN_BODY_TIMED_OUT` on that path.
- [ ] Existing single-wave-hang behavior unchanged.
- [ ] `Resource._shutdown_children()`'s existing single-batch behavior is unchanged (default `waves_left=1` reproduces the current formula exactly) — add a test pinning this if none exists.
- [ ] `uv run pytest tests/unit/core tests/integration -n 4` passes.
- [ ] `prek -a` and `prek pyright -a --stage pre-push` are clean.

Surfaced by the 2026-09-02 backend-core correctness audit (`design/audits/2026-09-02-backend-core-correctness-audit.md`).

### PR DESCRIPTION
**Pull request by NodeJSmith**
## Summary

`Hassette._shutdown_children()` loops through dependency-ordered waves, calling `children_budget_remaining()` for each wave's timeout. The function floors the result at `CHILDREN_FLOOR_SECONDS` (1.0s) — but that floor applies independently per wave. With W hung waves, the body overruns `body_deadline` by up to W × 1.0s while the coordinator margin is only 3.0s (at the 30s default). At ~7 waves, 7.0s > 3.0s — the coordinator's outer backstop fires first, making `SHUTDOWN_BODY_TIMED_OUT` the normal outcome in the exact scenario the graceful wave-by-wave logic was designed for.

## Changes

**`src/hassette/resources/lifecycle.py`:**
- Add `CHILDREN_WAVE_FLOOR_SECONDS = 0.1` — a much smaller per-wave floor for multi-wave shutdowns.
- `children_budget_remaining()` gains a `waves_left` keyword parameter (default `1`). When `waves_left > 1`, the remaining time is divided across waves and floored at `CHILDREN_WAVE_FLOOR_SECONDS`. When `waves_left=1` (the default and the final wave), the existing `CHILDREN_FLOOR_SECONDS` floor applies — giving the most-foundational resources (DB, sync executor) a bigger grace window. `Resource._shutdown_children()`'s single-batch behavior is byte-for-byte unchanged.

**`src/hassette/core/core.py`:**
- `Hassette._shutdown_children()` pre-computes non-empty waves, passes `waves_left` to `children_budget_remaining()`, and decrements each iteration.
- Worst-case total floor overrun with W waves is now `(W-1) * 0.1 + 1.0` = 1.6s at ~7 waves, well within the 3.0s coordinator margin.

## Testing

- `test_multi_wave_hang_finishes_within_coordinator_margin` — verifies that a 6-wave shutdown with all children hanging completes within body_deadline (i.e. the coordinator margin is not breached).
- `test_children_budget_remaining_multi_wave_divides_across_waves` — unit test for the new `waves_left` division logic.
- `test_children_budget_remaining_multi_wave_uses_wave_floor_when_over_budget` — unit test confirming the `CHILDREN_WAVE_FLOOR_SECONDS` floor applies when over budget.
- `test_children_budget_remaining_single_wave_returns_floor_when_over_budget` — existing single-wave behavior preserved.
- `test_children_budget_remaining_single_wave_benefits_from_slack` — existing slack behavior preserved.

Ran `tests/unit/core/test_core_coverage.py` and `tests/unit/resources/lifecycle/test_shutdown.py` locally — 70/70 passed. CI runs the full backend + system + e2e suites.

Closes #1809

### CODE REVIEW

**Comment by NodeJSmith** [comment]
@coderabbitai review

**Comment by NodeJSmith** [comment]
@coderabbitai review

File: src/hassette/resources/lifecycle.py [hunk 1]
```diff
@@ -567,17 +579,30 @@ def hooks_pool_remaining(resource: _LifecycleHostP) -> float:
...
     """
     resource = typing.cast("LifecycleMixin", resource)
     budget = resource._shutdown_budget
     if budget is None:
         return resource.hassette.config.lifecycle.resource_shutdown_timeout_seconds
-    return max(budget.children_floor_seconds, budget.body_deadline - asyncio.get_running_loop().time())
+    remaining = budget.body_deadline - asyncio.get_running_loop().time()
+    if waves_left <= 1:
+        return max(budget.children_floor_seconds, remaining)
+    return max(CHILDREN_WAVE_FLOOR_SECONDS, remaining / waves_left)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/resources/lifecycle.py:605]: Skipping — this is the exact problem the PR fixes. The `remaining / waves_left` division adjusts dynamically as earlier waves complete (waves_left decrements after each wave), so a fast wave that finishes early increases the share for subsequent waves. The equal-division concern applies to the *old* code (which used a flat floor per wave), not to this fix.

File: src/hassette/core/core.py [hunk 2]
```diff
@@ -768,23 +768,30 @@ async def _shutdown_children(self) -> TeardownReport:
...
         waves_start = asyncio.get_running_loop().time()
-        for wave_types in reversed(self._init_waves):
+        for wave_types in non_empty_waves:
             wave = [type_to_instance[t] for t in wave_types if t in type_to_instance]
-            if not wave:
-                continue
-            timeout = min(self.config.lifecycle.resource_shutdown_timeout_seconds, children_budget_remaining(self))
+            timeout = min(
+                self.config.lifecycle.resource_shutdown_timeout_seconds,
+                children_budget_remaining(self, waves_left=waves_left),
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/core/core.py:795]: Skipping — the final-wave floor (1.0s) is within the coordinator margin (3.0s at default 30s total). If the final wave's floor pushes past body_deadline, the outer `asyncio.timeout(body_timeout)` in `_shutdown_body()` fires the TOTAL_TIMEOUT/FORCED_TERMINAL fallback, which force-finalizes all children. The per-wave evidence loss is only possible when the body budget is already nearly exhausted (earlier waves consumed it), and the outer timeout is the designed safety net for exactly that case.

File: src/hassette/resources/lifecycle.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/resources/lifecycle.py:605]: Skipping — YAGNI. The finding itself calculates 262 waves are needed to exceed the body budget, which is not a realistic scenario (production hassette has ~7 waves). The worst case at 7 waves is (6 * 0.1) + 1.0 = 1.6s, well within the 3.0s coordinator margin. The test `test_children_budget_remaining_multi_wave_uses_wave_floor_when_over_budget` already asserts this invariant.
